### PR TITLE
Suggest `fn` if `fun`, `func`, `function` or `def` is used to define a function

### LINKED ISSUE #99751: Suggest ```fn``` if ```fun```, ```func```, ```function``` or ```def``` is used instead
Given the following code: 

```rust
func foo() -> i32 {
    123
}

function foo() -> i32 {
    123
}
```

The current output is:

```
   Compiling playground v0.0.1 (/playground)
error: expected one of `!` or `::`, found `foo`
 --> src/main.rs:1:6
  |
1 | func foo() -> i32 {
  |      ^^^ expected one of `!` or `::`

error: could not compile `playground` due to previous error
```

Ideally the error message should suggest replacing ```func``` or ```function``` with ```fn```.

### PR DESCRIPTION
**Pull request by bindsdev**
Closes #99751

### CODE REVIEW

**Comment by akabinds** [comment]
@lcnr: According to the CI failure message, and a couple bot messages above, it seems like I somehow modified `miri`. To make the CI pass and hopefully unblock the merge, what would I do to prevent the CI from failing?

File: compiler/rustc_parse/src/parser/diagnostics.rs [hunk 1]
```diff
@@ -613,6 +613,17 @@ impl<'a> Parser<'a> {
             }
         }
 
+        if let TokenKind::Ident(symbol, _) = &self.prev_token.kind {
+            if ["def", "fun", "func", "function"].contains(&symbol.as_str()) {
+                err.span_suggestion_short(
+                    self.prev_token.span,
+                    &format!("write `fn` instead of `{symbol}` to declare a function"),
```
Comment (AgentUkulele) [inline compiler/rustc_parse/src/parser/diagnostics.rs:620]: Is this missing the `symbol` format?

Comment (akabinds) [reply compiler/rustc_parse/src/parser/diagnostics.rs:620]: I am a bit confused. What do you mean?

Comment (akabinds) [reply compiler/rustc_parse/src/parser/diagnostics.rs:620]: If you were expecting something like `&format!("stuff `{}` more stuff", symbol)`, you can embed identifiers inside `{}`.

Comment (AgentUkulele) [reply compiler/rustc_parse/src/parser/diagnostics.rs:620]: Oh I'm an idiot, my bad

**Comment by lcnr** [comment]
you've committed a submodule change. I don't know enough about `git` to give you the best way to do it, but it should be easiest to use `git reset HEAD~1` to undo your commit and then add all files while being careful to not readd the change to the `miri` submodule.

If you're still stuck I suggest asking for help with this in #compiler/help on zulip

**Comment by akabinds** [comment]
> you've committed a submodule change. I don't know enough about `git` to give you the best way to do it, but it should be easiest to use `git reset HEAD~1` to undo your commit and then add all files while being careful to not readd the change to the `miri` submodule.
> 
> 
> 
> If you're still stuck I suggest asking for help with this in #compiler/help on zulip

This did work, but it seems like I somehow merged the branches again, causing the reset to become void 🫤. I will resolve this when I return home in the evening.

**Comment by akabinds** [comment]
> you've committed a submodule change. I don't know enough about `git` to give you the best way to do it, but it should be easiest to use `git reset HEAD~1` to undo your commit and then add all files while being careful to not readd the change to the `miri` submodule.
> 
> 
> 
> If you're still stuck I suggest asking for help with this in #compiler/help on zulip

That did work, but it seems like I accidentally merged the branch, keeping Miri changes there. I will fix when I get home this evening.

**Comment by lcnr** [comment]
please merge your commits, this can be done using
```rust
git rebase HEAD~4 -i
```

and then changing `pick` for all but the first commit to `fixup` or just `f`.

I don't know this code enough to tell whether this would also emit this suggestion at places where it doesn't make sense, so

r? rust-lang/compiler

File: compiler/rustc_parse/src/parser/diagnostics.rs [hunk 2]
```diff
@@ -613,6 +613,17 @@ impl<'a> Parser<'a> {
             }
         }
 
+        if let TokenKind::Ident(symbol, _) = &self.prev_token.kind {
```
Comment (davidtwco) [inline compiler/rustc_parse/src/parser/diagnostics.rs:616]: Could you just include this in the `if let` that immediately preceeds it?

**Review by davidtwco — CHANGES_REQUESTED** [review]
This looks good, left one comment, but you'll need to remove the merge commits and reverts and things like that, for this change, you should have only one commit with the changes you're making (if you end up with conflicts, rebase instead of merging).

**Comment by davidtwco** [comment]
This still isn't right. I'd suggest just deleting your branch locally, creating a new identically-named branch from `master`, applying your change again, commiting and force-pushing. That's probably going to be easier.

**Comment by akabinds** [comment]
@davidtwco was that supposed to happen? I have never done a force push before and something feels wrong.

**Comment by davidtwco** [comment]
> @davidtwco was that supposed to happen? I have never done a force push before and something feels wrong.

I think it's just that you force pushed `master` to your branch, rather than `master` plus additional changes, and that made GitHub think it was merged.

**Comment by akabinds** [comment]
> > @davidtwco was that supposed to happen? I have never done a force push before and something feels wrong.
> 
> I think it's just that you force pushed `master` to your branch, rather than `master` plus additional changes, and that made GitHub think it was merged.

I see. It seems like no changes were recorded also. So what shall I do now?

**Comment by davidtwco** [comment]
> I see. It seems like no changes were recorded also. So what shall I do now?

I think you should be able to make the changes for this fix and then commit and push it.

**Comment by akabinds** [comment]
> > I see. It seems like no changes were recorded also. So what shall I do now?
> 
> I think you should be able to make the changes for this fix and then commit and push it.

Will I have to make another PR?

**Comment by davidtwco** [comment]
I don't think so, if you add changes this should re-open, but there's no harm in another PR if that's easier too.

**Comment by akabinds** [comment]
> I don't think so, if you add changes this should re-open, but there's no harm in another PR if that's easier too.

I pushed the changes about 14 hours ago. Seems like it wasn't reopened.

**Comment by davidtwco** [comment]
> > I don't think so, if you add changes this should re-open, but there's no harm in another PR if that's easier too.
> 
> I pushed the changes about 14 hours ago. Seems like it wasn't reopened.

Feel free to open a new PR then, and include `r? @davidtwco` (outside of a code block) in the description so that I'm assigned.

**Pull request merged**